### DISCUSSION TITLE
Advice for microservice architecture and shared Prisma Client wrapper

### DISCUSSION THREAD

**Discussion by jpapini** [prisma/prisma#13120] [Q&A] [answered]
Hello everyone!

I'm writing to you today because I need some architecture advice for my microservices project using Prisma and Nest.js.
I'll try to be pretty clear in my explanations, hoping you'll be able to understand.

Let's say I have 2 services `service-a` and `service-b` both using 2 databases and 2 Prisma disctint schemas.

My Prisma client is generated in a dedicated `generated/client` folder for each services with this configuration:

```prisma
generator client {
    provider = "prisma-client-js"
    output   = "../generated/prisma-client"
}
```

And with a configured TypeScript, I can import my client like this:
`import { PrismaClient } from '~/prisma-client';`

Knowing now that I'm working with Nest.js, I need to make a wrapper around the PrismaClient to implement some business logic and to make it injectable in the Nest.js IoC (using the following tutorial [here](https://docs.nestjs.com/recipes/prisma#use-prisma-client-in-your-nestjs-services)).

```typescript
import type { INestApplication, OnModuleInit } from '@nestjs/common';

import { PrismaClient } from '~/prisma-client';

export class DBClient extends PrismaClient implements OnModuleInit
{
    async onModuleInit() {
        await this.$connect();
    }

    async enableShutdownHooks(app: INestApplication) {
        this.$on('beforeExit', async () => {
            await app.close();
        });
    }

    // Lot of more logics here
}

```

Knowing that I work in a monorepo with PNPM, we have this kind of architecture:

```
.
└── services/
    ├── service-a/
    │   ├── generated/
    │   │   └── client/
    │   │       └── ...
    │   ├── prisma/
    │   │   └── schema.prisma
    │   └── src/
    │       ├── db-client/
    │       │   └── db-client.ts
    │       └── ...
    └── service-b/
        ├── generated/
        │   └── client/
        │       └── ...
        ├── prisma/
        │   └── schema.prisma
        └── src/
            ├── db-client/
            │   └── db-client.ts
            └── ...
```

My question now:

Let's imagine now that I have more than a dozen microservices all using the same code regarding my wrapper to Prisma Client.
I'd like to create a `lib-database` and put my wrapper to Prisma Client in it so I don't duplicate code unnecessarily. Especially if later I want to add shared logic between my services.
The difficulty is that my `DBClient` wrapper needs PrismaClient and its TypeScript definitions in order to compile this lib.

My first steps were to declare `@prisma/client` as a `peerDependencies` of my `lib-database` in order to directly use the client used in my services.
But without success because the typing of the PrismaClient class is not available until the client is generated (and I can't generate the PrismaClient types in my `lib-database` because it is supposed to be generic for all my microservices).

Have you ever faced this kind of situation? Do you have any idea how to architect my project to share this wrapper across all my microservices?

Thanks in advance for your advice!

**Comment by nurul3101** [comment, marked as answer]
Hey @jpapini 👋 

Here's an example by our developer advocate of how you could share your types in a monorepo setup which would be helpful to you: [monorepo-prisma](https://github.com/nikolasburk/monorepo-prisma)

Let me know if this reference solves the issue for you. 

The idea is that If you’re using a monorepo, you can directly share the backend types by importing them in your `lib-database` package.
If they are in different repos, it's best to create an NPM package that includes the types which you can install on the other app and that can be imported.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
